File: ossium/transfer_function.py
import struct
import os

class TransferFunction:
    MAGIC_NUMBER = b'TF01'  # Magic number to identify our transfer function file format
    
    def __init__(self, filepath):
        self.filepath = filepath  # Store full path privately
        self.filename = os.path.basename(filepath)  # Store just the filename for the API
        self.load(filepath)

    def __getstate__(self):
        """Return a JSON-serializable state dictionary"""
        state = self.__dict__.copy()
        # Remove private attributes from serialization
        state.pop('_data', None)
        state.pop('filepath', None)
        return state

    def load(self, filepath):
        """Load transfer function from a single .tf file"""
        with open(filepath, 'rb') as f:
            # Verify magic number
            magic = f.read(4)
            if magic != self.MAGIC_NUMBER:
                raise ValueError("Invalid transfer function file format")
            
            self.n_colours = struct.unpack('I', f.read(4))[0]
            self._data = f.read()

    def save(self, output_filename):
        """Create a new transfer function file"""
        with open(output_filename, 'wb') as f:
            # Write magic number
            f.write(TransferFunction.MAGIC_NUMBER)
            
            # Write header
            f.write(struct.pack('I', self.n_colours))
            
            # Write color data
            f.write(self._data)

    def stream_data(self):
        """Stream transfer function data."""
        # Skip header (magic number + n_colours = 8 bytes)
        yield self._data

File: src/transfer_function.ts
export class TransferFunction {
    // Magic number to identify the transfer function file format
    public static readonly MAGIC_NUMBER = 'TF01';

    private _name: string;
    private _size: number;
    private _data: ArrayBuffer;
    
    constructor(name?: string) {
        this._name = name || '';
        this._size = 1;
        this._data = new ArrayBuffer(0);
    }

    public get name(): string { return this._name; }
    public get size(): number { return this._size; }
    public get data(): ArrayBuffer { return this._data; }

    public async load(file: File): Promise<void> {
        const buffer = await file.arrayBuffer();
        const view = new DataView(buffer);
        
        // Check magic number 'TF01'
        const magic = new TextDecoder().decode(buffer.slice(0, 4));
        if (magic !== TransferFunction.MAGIC_NUMBER) {
            throw new Error('Invalid transfer function');
        }

        // Read size of transfer function (4 bytes after magic number)
        this._size = view.getUint32(4, true);
        
        // Store the remaining data (color data)
        this._data = buffer.slice(8); // Skip magic number and size
    }
}